### PR TITLE
fs: Fix Shutdown() behavior and handle tests properly.

### LINKED ISSUE #4795: TestFSGetBucketInfo fails sporadically on appveyor
```
--- FAIL: TestFSGetBucketInfo (1.69s)
	fs-v1_test.go:119: BucketNotFound error not returned
```

## Expected Behavior

Unit test must pass always

## Current Behavior

Unit test fails sporadically

## Possible Solution

The unit test uses `removeAll(disk)` to simulate `errDiskNotFound`. Minio holds a read-write lock on `format.json` file and a remove fs directory may fail. Alternately, we could test like,
```go
// change FS object layer disk path to point to a non-existent directory
newFS := fs
newDisk := filepath.Join(globalTestTmpDir, "minio-"+nextSuffix())
newFS.fsPath = newDisk
if err := newFS.GetBucketInfo(bucket); err != BucketNotFound {
   ...
}
```
## Steps to Reproduce (for bugs)

1. make test
2.
3.
4.

## Your Environment

* Version used (`minio version`):
* Environment name and version (e.g. nginx 1.9.1):
* Server type and version:
* Operating System and version (`uname -a`):
* Link to your project:

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
Fix Shutdown() behavior and handle tests properly.

## Motivation and Context
Fixes #4795

## How Has This Been Tested?
Manually

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: cmd/fs-v1_test.go [hunk 1]
```diff
@@ -265,7 +266,7 @@ func TestFSDeleteBucket(t *testing.T) {
 	obj.MakeBucketWithLocation(bucketName, "")
 
 	// Delete bucker should get error disk not found.
```
Comment (krisis) [inline cmd/fs-v1_test.go]: unrelated typo, yet close enough to change.
s/bucker/bucket

File: cmd/fs-v1-multipart_test.go [hunk 2]
```diff
@@ -20,15 +20,76 @@ import (
...
+	close(fs.staleMultipartUploadsDoneCh)
+
+	// Allocate a new done channel.
+	fs.staleMultipartUploadsDoneCh = make(chan struct{})
+
+	go fs.cleanupStaleMultipartUploadsRoutine(20*time.Millisecond, 0)
+
+	// Wait for 200ms such that we have given enough time for
+	// cleanup routine to kick in.
+	time.Sleep(40 * time.Millisecond)
```
Comment (krisis) [inline cmd/fs-v1-multipart_test.go]: comment and code are not in sync. Either the comment should be 40ms or Sleep should take 200ms.

File: cmd/fs-v1-multipart_test.go [hunk 3]
```diff
@@ -20,15 +20,76 @@ import (
...
+	objectName := "object"
+
+	obj.MakeBucketWithLocation(bucketName, "")
+	_, err := obj.NewMultipartUpload(bucketName, objectName, nil)
+	if err != nil {
+		t.Fatal("Unexpected err: ", err)
+	}
+
+	// Close the current active go-routine.
+	close(fs.staleMultipartUploadsDoneCh)
```
Comment (krisis) [inline cmd/fs-v1-multipart_test.go]: Close the (active) go-routine before NewMultipartUpload request  to ensure that the new routine is the one that performed the clean up.

File: cmd/fs-v1-multipart_test.go [hunk 4]
```diff
@@ -20,15 +20,76 @@ import (
...
+	fs.staleMultipartUploadsDoneCh = make(chan struct{})
+
+	go fs.cleanupStaleMultipartUploadsRoutine(20*time.Millisecond, 0)
+
+	// Wait for 200ms such that we have given enough time for
+	// cleanup routine to kick in.
+	time.Sleep(40 * time.Millisecond)
+
+	// Close the current active go-routine.
+	close(fs.staleMultipartUploadsDoneCh)
```
Comment (krisis) [inline cmd/fs-v1-multipart_test.go]: we should add a check that verifies that the clean up happened.

File: cmd/fs-v1-multipart.go [hunk 5]
```diff
@@ -973,19 +974,22 @@ func (fs fsObjects) cleanupStaleMultipartUpload(bucket string) (err error) {
 }
 
 // Remove multipart uploads left unattended for more than `fsMultipartExpiry` seconds.
```
Comment (krisis) [inline cmd/fs-v1-multipart.go]: comment should be updated to indicate use of multipartExpiry argument.

File: cmd/fs-v1-multipart.go [hunk 6]
```diff
@@ -938,27 +938,28 @@ func (fs fsObjects) AbortMultipartUpload(bucket, object, uploadID string) error
 }
 
 // Remove multipart uploads left unattended in a given bucket older than `fsMultipartExpiry`
```
Comment (krisis) [inline cmd/fs-v1-multipart.go]: comment needs to be updated to use `expiry` argument.

File: cmd/fs-v1-multipart_test.go [hunk 7]
```diff
@@ -20,15 +20,76 @@ import (
...
+	bucketName := "bucket"
+	objectName := "object"
+
+	obj.MakeBucketWithLocation(bucketName, "")
+	_, err := obj.NewMultipartUpload(bucketName, objectName, nil)
+	if err != nil {
+		t.Fatal("Unexpected err: ", err)
+	}
+
+	// Close the current active go-routine.
```
Comment (krisis) [inline cmd/fs-v1-multipart_test.go]: this should happen before creating a new multipart upload to ensure that the following cleanup is the one removing 'stale' uploads

File: cmd/fs-v1-multipart_test.go [hunk 8]
```diff
@@ -20,15 +20,76 @@ import (
...
+	close(fs.staleMultipartUploadsDoneCh)
+
+	if err = fs.cleanupStaleMultipartUpload(bucketName, 0); err != nil {
+		t.Fatal("Unexpected err: ", err)
+	}
```
Comment (krisis) [inline cmd/fs-v1-multipart_test.go]: need to check that the uploadId directory is removed as expected.

**Review by krisis — CHANGES_REQUESTED** [review]

File: cmd/fs-v1-multipart.go [hunk 9]
```diff
@@ -938,27 +938,28 @@ func (fs fsObjects) AbortMultipartUpload(bucket, object, uploadID string) error
...
 	var lmi ListMultipartsInfo
 
 	for {
 		// List multipart uploads in a bucket 1000 at a time
 		prefix := ""
 		lmi, err = fs.listMultipartUploadsHelper(bucket, prefix, lmi.KeyMarker, lmi.UploadIDMarker, slashSeparator, 1000)
 		if err != nil {
-			errorIf(err, fmt.Sprintf("Failed to list uploads of %s for cleaning up of multipart uploads older than %d weeks", bucket, fsMultipartExpiry))
+			errorIf(err, "Failed to list uploads of %s for cleaning up of multipart uploads older than %d weeks",
+				bucket, fsMultipartExpiry)
```
Comment (balamurugana) [inline cmd/fs-v1-multipart.go]: This error message is not clear or too verbose.

File: cmd/fs-v1-multipart.go [hunk 10]
```diff
@@ -938,27 +938,28 @@ func (fs fsObjects) AbortMultipartUpload(bucket, object, uploadID string) error
...
 
-		// Remove uploads (and its parts) older than 2 weeks
+		// Remove uploads (and its parts) older than expiry duration.
 		for _, upload := range lmi.Uploads {
 			uploadIDPath := pathJoin(fs.fsPath, minioMetaMultipartBucket, bucket, upload.Object, upload.UploadID)
-			st, err := fsStatDir(uploadIDPath)
-			if err != nil {
-				errorIf(err, "Failed to stat uploads directory", uploadIDPath)
-				return err
+			var st os.FileInfo
```
Comment (balamurugana) [inline cmd/fs-v1-multipart.go]: you could move out of `st` from `for` loop

File: cmd/fs-v1-multipart.go [hunk 11]
```diff
@@ -973,19 +974,22 @@ func (fs fsObjects) cleanupStaleMultipartUpload(bucket string) (err error) {
...
-		bucketInfos, err := fs.ListBuckets()
-		if err != nil {
-			errorIf(err, fmt.Sprintf("Failed to list buckets for cleaning up of multipart uploads older than %d weeks", fsMultipartExpiry))
+		select {
+		case <-fs.staleMultipartUploadsDoneCh:
 			return
+		case <-timer.C:
+			bucketInfos, err := fs.ListBuckets()
+			if err != nil {
+				errorIf(err, "Failed to list buckets for cleaning up of multipart uploads older than %d weeks",
```
Comment (balamurugana) [inline cmd/fs-v1-multipart.go]: This error message is too verbose.

**Review by balamurugana — CHANGES_REQUESTED** [review]

**Review by krisis — APPROVED** [review]

**Review by balamurugana — APPROVED** [review]

**Pull request merged**